src/enable_variable_ops/ngraph_capture_variables.cc

```diff
@@ -0,0 +1,285 @@
+                       << replacement->DebugString();
+
+        // TODO(Mingshan): Variable doesn't have input edges, need to confirm
+        // Though edges will be removed when we remove the node
+        // we specifically remove the edges to be sure
+        for (auto edge : node->in_edges()) {
+          NGRAPH_VLOG(4) << "Replacing: " << edge->DebugString();
+          graph->AddEdge(edge->src(), edge->src_output(), replacement,
+                         edge->dst_input());
+          graph->RemoveEdge(edge);
+        }
+
+        for (auto edge : node->out_edges()) {
+          edges.push_back(edge);
+        }
+
+        for (auto edge : edges) {
+          NGRAPH_VLOG(4) << "Replacing: " << edge->DebugString();
+          graph->AddEdge(replacement, edge->src_output(), edge->dst(),
+                         edge->dst_input());
+          graph->RemoveEdge(edge);
+        }
+
+        replaced_nodes.push_back(node);
+      }
+
+      else if (IsTFAssignType(node->type_string())) {
+        NGRAPH_VLOG(1) << "Capturing: " << node->name();
+
+        auto node_type = node->type_string();
+        auto node_replacement_type = GetNGAssignType(node_type);
+        NGRAPH_VLOG(1) << "Node Type: " << node_type
+                       << " , Node Replacement Type " << node_replacement_type;
+
+        DataType dtype;
+        TF_RETURN_IF_ERROR(GetNodeAttr(node->attrs(), "T", &dtype));
+        Node* replacement;
+
+        NodeBuilder::NodeOut input_ref;
+        NodeBuilder::NodeOut input_val;
+
+        for (auto edge : node->in_edges()) {
+          if (edge == NULL) {
+            NGRAPH_VLOG(1) << "Capturing " << node_type
+                           << " op, but found null edge: ";
+            continue;
+          }
+
+          if (edge->dst()->IsOp() && !edge->IsControlEdge() &&
+              IsRefType(edge->dst()->input_type(edge->dst_input()))) {
+            input_ref = NodeBuilder::NodeOut(edge->src(), edge->src_output());
+          } else {
+            input_val = NodeBuilder::NodeOut(edge->src(), edge->src_output());
+          }
+        }
+
+        // TODO(amprocte): Do we need to copy "_" attributes?
+        TF_RETURN_IF_ERROR(NodeBuilder(node->name(), node_replacement_type)
+                               .Attr("validate_shape", true)
+                               .Attr("use_locking", true)
+                               .Attr("T", dtype)
+                               .Attr("ngraph_graph_id", 0)
+                               .Input(input_ref)
+                               .Input(input_val)
+                               .Device(node->assigned_device_name())
+                               .Finalize(graph, &replacement));
+
+        NGRAPH_VLOG(1) << "Successfully constructed " << node_replacement_type
+                       << " Node Def";
+
+        replacement->set_assigned_device_name(node->assigned_device_name());
+
+        std::vector<const Edge*> edges;
```

Comment:
Lines 161 - 192 where we replace the edges are similar to lines 84-112, 246-268

Maybe we can make them common for all cases and move them outside the if-else ladder

Reply:
This requires some refactoring. I can create a function and see if that works. Working on this.